File: algorithms/algorithms-cplusplus/leetcode/Question_0046_Permutations.cc
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> permute(vector<int>& nums) {
    vector<vector<int>> res;
    if (nums.empty()) {
      return res;
    }
    help(nums, 0, res);
    return res;
  }

  void help(vector<int>& nums, int i, vector<vector<int>>& res) {
    if (i >= nums.size()) {
      res.push_back(nums);
      return;
    }
    for (int j = i; j < nums.size(); j++) {
      swap(nums[i], nums[j]);
      help(nums, i + 1, res);
      swap(nums[i], nums[j]);
    }
  }
};

File: algorithms/algorithms-python/leetcode/Question_0046_Permutations.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        if not nums:
            return res
        self.help(nums, 0, res)
        return res

    def help(self, nums, i, res):
        if i == len(nums) - 1:
            res.append(nums[:])
            return

        for j in range(i, len(nums)):
            nums[i], nums[j] = nums[j], nums[i]
            self.help(nums, i + 1, res)
            nums[i], nums[j] = nums[j], nums[i]